### PR TITLE
V2 optimizations

### PR DESCRIPTION
**Pull request by fabrice-quadrata**
## Description
This Pull Request is refactoring how attributes are stored and can be queried for optimization.

Notable changes:
- How we store attributes
- How we query attributes
- Adding new functions to query attributes in bulk
- Modify payments to issuers/protocols to issue event receipts
- Removing ERC20 payment support
- Removing Exclude/Include functions for issuers
- Remove OZ ERC1155 and replace it with custom QuadSoulbound

## Checklist
1. [x] Unit Testing?
2. [x] Backward Compatible?
3. [x] Upgrade Safe (storage)?
4. [x] Updated Confluence?
5. [x] Updated Gitbook?

### CODE REVIEW

File: contracts/storage/QuadConstant.sol [hunk 1]
```diff
@@ -0,0 +1,17 @@
+//SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.16;
+
+contract QuadConstant {
+    bytes32 public constant ISSUER_ROLE = keccak256("ISSUER_ROLE");
+    bytes32 public constant GOVERNANCE_ROLE = keccak256("GOVERNANCE_ROLE");
+    bytes32 public constant READER_ROLE = keccak256("READER_ROLE");
+    bytes32 public constant PAUSER_ROLE = keccak256("PAUSER_ROLE");
+
+    bytes32 public constant DIGEST_TO_SIGN = 0x37937bf5ff1ecbf00bbd389ab7ca9a190d7e8c0a084b2893ece7923be1d2ec85;
```
Comment (faze-quadrata) [inline contracts/storage/QuadConstant.sol:10]: 👍 
>>> Web3.keccak(text=MESSAGE_TO_SIGN).hex()
'0x37937bf5ff1ecbf00bbd389ab7ca9a190d7e8c0a084b2893ece7923be1d2ec85'

File: contracts/QuadPassport.sol [hunk 2]
```diff
@@ -1,340 +1,362 @@
...
-
-    fallback() external payable {}
-
-    /// @notice Claim and mint a wallet account Quadrata Passport
+    /// @notice Set attributes for a Quadrata Passport (Only Individuals)
+    /// @notice If passing a `DID` in _config.attrTypes, make sure that it's always first in the list
     /// @dev Only when authorized by an eligible issuer
-    /// @param _config Input paramters required to mint
+    /// @param _config Input paramters required to set attributes
     /// @param _sigIssuer ECDSA signature computed by an eligible issuer to authorize the mint
```
Comment (teddy-quadrata) [inline contracts/QuadPassport.sol:39]: Update Comment: 
`to authorize the update`

File: contracts/QuadPassport.sol [hunk 3]
```diff
@@ -1,340 +1,362 @@
...
+            if(_did == bytes32(0)){
+                Attribute[] memory dIDAttrs = _attributes[keccak256(abi.encode(_account, ATTRIBUTE_DID))];
+                require(dIDAttrs.length > 0 && dIDAttrs[0].value != bytes32(0), "MISSING_DID");
+                _did = dIDAttrs[0].value;
+
+            }
+            return keccak256(abi.encode(_did, _attribute));
         }
+
+        require(false, "ATTRIBUTE_NOT_ELIGIBLE");
```
Comment (teddy-quadrata) [inline contracts/QuadPassport.sol:214]: Should be deleted, no?

**Review by teddy-quadrata — CHANGES_REQUESTED** [review]

File: contracts/QuadPassport.sol [hunk 4]
```diff
@@ -1,340 +1,362 @@
...
```
Comment (teddy-quadrata) [inline contracts/QuadPassport.sol:214]:
```suggestion
        revert("ATTRIBUTE_NOT_ELIGIBLE");
```

**Review by teddy-quadrata — APPROVED** [review]

**Pull request merged**